### PR TITLE
Add Type Annotations to pl_bolts/utils/*

### PR DESCRIPTION
**Pull request by hassiahk**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning-bolts/issues/434. This PR adds type annotations to `pl_bolts.utils.*`

## Before submitting
- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [X] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [X] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by hassiahk** [comment]
@akihironitta, need help with the below two errors. I am not sure how to fix them.
- `Cannot access "__init__" directly`
- `Module has no attribute "_empty"`

File: pl_bolts/utils/semi_supervised.py [hunk 1]
```diff
@@ -51,19 +53,19 @@ def balance_classes(X: np.ndarray, Y: list, batch_size: int):
...
-    final_batches_y = [[] for i in range(nb_batches)]
+    final_batches_x: List[Any] = [[] for i in range(nb_batches)]
+    final_batches_y: List[Any] = [[] for i in range(nb_batches)]
 
     # Y needs to be np arr
     Y = np.asarray(Y)
 
     # pick chunk size for each class using the largest split
-    chunk_size = []
+    chunk_size_class = []
```
Comment (hassiahk) [inline pl_bolts/utils/semi_supervised.py:63]: @akihironitta, `mypy` was giving error if I don't change `chunk_size` to something else because it is `List[int]` and when we do `chunk_size = max(chunk_size)` it is becoming `int`. Let me know if this okay to do.

File: pl_bolts/utils/semi_supervised.py [hunk 2]
```diff
@@ -51,19 +53,19 @@ def balance_classes(X: np.ndarray, Y: list, batch_size: int):
     nb_batches = math.ceil(len(Y) / batch_size)
 
     # sort by classes
-    final_batches_x = [[] for i in range(nb_batches)]
-    final_batches_y = [[] for i in range(nb_batches)]
+    final_batches_x: List[Any] = [[] for i in range(nb_batches)]
+    final_batches_y: List[Any] = [[] for i in range(nb_batches)]
```
Comment (hassiahk) [inline pl_bolts/utils/semi_supervised.py:57]: `mypy` was asking me to annotate these two.

Comment (akihironitta) [inline pl_bolts/utils/semi_supervised.py:57]: In my opinion, we shouldn't use `Any` where possible, so can we rewrite it as follows?
```suggestion
    final_batches_x: List[List[int]] = [[] for i in range(nb_batches)]
    final_batches_y: List[List[int]] = [[] for i in range(nb_batches)]
```

File: pl_bolts/utils/semi_supervised.py (hunk 1 above)
Comment (akihironitta) [reply pl_bolts/utils/semi_supervised.py:63]: I think you're right. How about renaming the variable `chunk_size_class` to `chunk_sizes`?

File: pl_bolts/utils/arguments.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ class LightningArgumentParser(ArgumentParser):
         # args.data -> data args
         # args.model -> model args
     """
-    def __init__(self, *args, ignore_required_init_args=True, **kwargs):
+    def __init__(self, *args: Any, ignore_required_init_args: bool = True, **kwargs: Any):
```
Comment (akihironitta) [inline pl_bolts/utils/arguments.py:35]:
```suggestion
    def __init__(self, *args: Any, ignore_required_init_args: bool = True, **kwargs: Any) -> None:
```

File: pl_bolts/utils/arguments.py [hunk 4]
```diff
@@ -41,10 +41,10 @@ def __init__(self, *args, ignore_required_init_args=True, **kwargs):
         super().__init__(*args, **kwargs)
         self.ignore_required_init_args = ignore_required_init_args
 
-        self._default_obj_args = dict()
-        self._added_arg_names = []
+        self._default_obj_args: Dict[str, List[LitArg]] = dict()
```
Comment (akihironitta) [inline pl_bolts/utils/arguments.py:44]: This is completely off topic and not a strong opinion, but we should use `{}` instead of `dict()` to instantiate an empty dictionary as I remember `{}` is faster than `dict()`.

@hassiahk Would you be interested in submitting a seperate PR replacing `dict()` with `{}` in the codebase?

File: pl_bolts/utils/arguments.py [hunk 5]
```diff
@@ -32,7 +32,7 @@ class LightningArgumentParser(ArgumentParser):
         # args.data -> data args
         # args.model -> model args
     """
-    def __init__(self, *args, ignore_required_init_args=True, **kwargs):
+    def __init__(self, *args: Any, ignore_required_init_args: bool = True, **kwargs: Any):
         """
         Args:
             ignore_required_init_args (bool, optional): Whether to include positional args when adding
```
Comment (akihironitta) [inline pl_bolts/utils/arguments.py:38]: As we have function annotations here, the types of args should be automatically reflected in the docs, so can you remove these from docstrings?
```suggestion
            ignore_required_init_args: Whether to include positional args when adding
```

File: pl_bolts/utils/semi_supervised.py [hunk 6]
```diff
@@ -24,14 +26,14 @@ class Identity(torch.nn.Module):
...
+    def __init__(self) -> None:
         super(Identity, self).__init__()
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Tensor:
         return x
 
 
-def balance_classes(X: np.ndarray, Y: list, batch_size: int):
+def balance_classes(X: np.ndarray, Y: list, batch_size: int) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (akihironitta) [inline pl_bolts/utils/semi_supervised.py:36]: May we be more specific here?
```suggestion
def balance_classes(X: np.ndarray, Y: List[int], batch_size: int) -> Tuple[np.ndarray, np.ndarray]:
```

File: pl_bolts/utils/warnings.py [hunk 7]
```diff
@@ -1,10 +1,16 @@
...
+from typing import Any, Callable
 
 MISSING_PACKAGE_WARNINGS = {}
 
 
-def warn_missing_pkg(pkg_name: str, pypi_name: str = None, extra_text: str = None, stdout_func=warnings.warn):
+def warn_missing_pkg(
+    pkg_name: str,
+    pypi_name: str = None,
+    extra_text: str = None,
```
Comment (akihironitta) [inline pl_bolts/utils/warnings.py:11]:
```suggestion
    pypi_name: Optional[str] = None,
    extra_text: Optional[str] = None,
```

**Review by akihironitta — CHANGES_REQUESTED** [review]
@hassiahk Thank you for your contribution! I left some comments on the changes. Could you have a look? As for your comment below, I'll look into it later...
> @akihironitta, need help with the below two errors. I am not sure how to fix them.
> - `Cannot access "__init__" directly`
> - `Module has no attribute "_empty"`

File: pl_bolts/utils/arguments.py (hunk 4 above)
Comment (hassiahk) [reply pl_bolts/utils/arguments.py:44]: Sure, I can do this.

File: pl_bolts/utils/arguments.py (hunk 5 above)
Comment (hassiahk) [reply pl_bolts/utils/arguments.py:38]: Sure, can do this as well.

File: pl_bolts/utils/semi_supervised.py (hunk 2 above)
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:57]: Yes, I was not sure regarding the type, so keep it as `Any`, will change that now.

File: pl_bolts/utils/semi_supervised.py (hunk 6 above)
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:36]: My bad, I think I missed this because it was already annotated.😄

Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:36]: Should I change it to `Sequence[int]` since it can be a `list` or a `tuple`?

File: pl_bolts/utils/semi_supervised.py (hunk 2 above)
Comment (akihironitta) [reply pl_bolts/utils/semi_supervised.py:57]: Sorry, what I've just mentioned is completely false statement.  Still, it would be nice to have a more specific type, but we can keep it as is since it seems not obvious...

Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:57]: Okay, sure.

**Comment by hassiahk** [comment]
@akihironitta, I took care of your requested changes but I also had to change `Y -> labels` in `balance_classes` while annotating it as `Sequence[int]` or `List[int]` because `mypy` is throwing below errors:
- `Argument 1 to "len" has incompatible type "int"; expected "Sized"` -> `mypy` is basically saying that you did not take care of `Y` being `None`.
- `Value of type "int" is not indexable` -> Not sure of this.

Another workaround would be to annotate `Y` as plain `Sequence` or `List`. Let me know your thoughts.

File: pl_bolts/utils/arguments.py (hunk 4 above)
Comment (hassiahk) [reply pl_bolts/utils/arguments.py:44]: @akihironitta, the entire codebase or just the `pl_bolts.utils.*`?

Comment (akihironitta) [reply pl_bolts/utils/arguments.py:44]: @hassiahk  Please do to the whole codebase in a single PR :]

File: pl_bolts/utils/semi_supervised.py (hunk 6 above)
Comment (akihironitta) [reply pl_bolts/utils/semi_supervised.py]: @hassiahk On second thought, I don't think the annotated types are not correct. In the tests, we use the function as the following:
```python
tests/utils/test_semi_supervised.py:9:def test_balance_classes():
tests/utils/test_semi_supervised.py-10-    X = torch.rand(100, 3, 32, 32)
tests/utils/test_semi_supervised.py-11-    c1 = torch.zeros(20, 1)
tests/utils/test_semi_supervised.py-12-    c2 = torch.zeros(20, 1) + 1
tests/utils/test_semi_supervised.py-13-    c3 = torch.zeros(60, 1) + 2
tests/utils/test_semi_supervised.py-14-
tests/utils/test_semi_supervised.py-15-    y = torch.cat([c1, c2, c3], dim=0).int().cpu().numpy().flatten().tolist()
tests/utils/test_semi_supervised.py:16:    (balanced_X, balanced_Y) = balance_classes(X, y, batch_size=10)
```
which suggests we should annotate them differently from the current ones.

**Review by akihironitta — CHANGES_REQUESTED** [review]
@hassiahk Thank you very much for your work!

For `Module has no attribute "_empty"`, let's replace `inspect._empty` with `inspect.Parameter.empty` because both should be equivalent which is defined in [`inspect`](https://github.com/python/cpython/blob/master/Lib/inspect.py). The same issue was also discussed in https://github.com/allenai/allennlp/pull/1191#discussion_r201096345.

For `Cannot access "__init__" directly`, I couldn't find any existing solution, so shall we just skip the line with `# type: ignore`...?

File: pl_bolts/utils/semi_supervised.py [hunk 8]
```diff
@@ -24,46 +26,46 @@ class Identity(torch.nn.Module):
...
+    def __init__(self) -> None:
         super(Identity, self).__init__()
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Tensor:
         return x
 
 
-def balance_classes(X: np.ndarray, Y: list, batch_size: int):
+def balance_classes(X: np.ndarray, labels: Sequence[int], batch_size: int) -> Tuple[np.ndarray, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:36]: Ah, I did not think it can be `torch.Tensor` or `numpy.ndarray`, I think this will solve the problem.
> (or you think of other types, too?).
It can also be `Tuple`, hence I kept it as `Sequence`

File: pl_bolts/utils/semi_supervised.py [hunk 9]
```diff
@@ -24,46 +26,46 @@ class Identity(torch.nn.Module):
...
     """
     Makes sure each batch has an equal amount of data from each class.
     Perfect balance
 
     Args:
         X: input features
-        Y: mixed labels (ints)
+        labels: mixed labels (ints)
```
[the remark this answers is not in the document]
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:43]: Okay, sure. I think annotating it as `Union` will solve the problem. If not, I can rename the variables that are inside the function.

File: pl_bolts/utils/semi_supervised.py (hunk 6 above)
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py]: Annotated `X` as `Union[Tensor, np.ndarray]` and `Y` as `Union[Tensor, np.ndarray, Sequence[int]]`.

File: pl_bolts/utils/semi_supervised.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py]: Annotating it as `Union` solved the problem.

**Comment by hassiahk** [comment]
@akihironitta, made requested changes. Kindly review it once again. Annotated `X` as `Union[Tensor, np.ndarray]` and `Y` as `Union[Tensor, np.ndarray, Sequence[int]]`.

File: pl_bolts/utils/semi_supervised.py [hunk 10]
```diff
@@ -24,14 +26,18 @@ class Identity(torch.nn.Module):
...
 
-    def forward(self, x):
+    def forward(self, x: Tensor) -> Tensor:
         return x
 
 
-def balance_classes(X: np.ndarray, Y: list, batch_size: int):
+def balance_classes(
+    X: Union[Tensor, np.ndarray],
+    Y: Union[Tensor, np.ndarray, Sequence[int]],
```
Comment (akihironitta) [inline pl_bolts/utils/semi_supervised.py:38]: I'm still concerned about if the annotated types include all the possibility, but I think these are fine for now.

File: pl_bolts/utils/semi_supervised.py [hunk 11]
```diff
@@ -51,19 +57,19 @@ def balance_classes(X: np.ndarray, Y: list, batch_size: int):
     nb_batches = math.ceil(len(Y) / batch_size)
 
     # sort by classes
-    final_batches_x = [[] for i in range(nb_batches)]
-    final_batches_y = [[] for i in range(nb_batches)]
+    final_batches_x: List[Any] = [[] for i in range(nb_batches)]
+    final_batches_y: List[Any] = [[] for i in range(nb_batches)]
```
Comment (akihironitta) [inline pl_bolts/utils/semi_supervised.py:61]: Since these variables are the output of the function, I think they should match with the return type of the function `balance_classes()`:
```python
) -> Tuple[np.ndarray, np.ndarray]:
```
but in order to do that, you might need to change/define other variables, too. We can tighten these types later in another PR, but it'd be nice to do it in this PR. Do you think you can improve typing here?

**Review by akihironitta — APPROVED** [review]
@hassiahk Thank you for your work! LGTM :]

File: pl_bolts/utils/arguments.py (hunk 4 above)
Comment (akihironitta) [reply pl_bolts/utils/arguments.py:44]: @hassiahk Look forward to reviewing your PR on this :]

File: pl_bolts/utils/semi_supervised.py (hunk 11 above)
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:61]: Sure, I will see what I can do. :)

Comment (Borda) [inline pl_bolts/utils/semi_supervised.py:61]:
```suggestion
    final_batches_x: List[list] = [[] for i in range(nb_batches)]
    final_batches_y: List[list] = [[] for i in range(nb_batches)]
```

**Review by Borda — APPROVED** [review]

File: pl_bolts/utils/semi_supervised.py (hunk 10 above)
Comment (hassiahk) [reply pl_bolts/utils/semi_supervised.py:38]: I will use this `balance_classes` and see if we can use other types as well.
